### ISSUE TITLE
research: capture Grok-1 expert behavior as teacher targets for neuromorphic distillation

### ISSUE THREAD

**Issue by rmems** [Spikenaut/grok-ozempic#84]
## Planning attribution

**GPT-5.6 (high)**

## Context

`grok-ozempic` is already running controlled real-weight Grok-1 expert/precision experiments. A separate neuromorphic research thread is exploring whether a small SNN can act as an always-on supervisor for hardware/agent state.

A tempting idea is to "put a Grok expert on FPGA," but a full frontier MoE expert is far beyond the storage/compute budget of the current Artix-7 target. The useful research question is therefore different:

> Can a narrow, measurable behavior of a Grok-1 expert or block be captured as teacher targets and distilled into a much smaller temporal/SNN student?

This issue is **exploratory and evidence-only**. It must not disturb the current quantization critical path or claim that a compressed expert is equivalent to Grok-1.

## Goal

Add a reproducible **teacher-behavior capture** path for selected Grok-1 components so downstream projects can train a small surrogate/SNN against recorded targets.

The first milestone is artifact generation only: no SNN dependency and no FPGA dependency inside `grok-ozempic`.

## Research questions

1. Which narrow expert/block signals are stable enough to serve as teacher targets?
2. How much behavior is preserved under the precision arms already studied in `grok-ozempic`?
3. Can a small surrogate predict those targets from a reduced input/state representation?
4. Does the target remain useful when transferred to a different temporal domain such as hardware supervisor state?

Question 4 is explicitly a later experiment; do not assume transfer will work.

## Phase 0 — choose one narrow target

Do not export arbitrary giant activation dumps.

Pick **one** initial target family, for example:

### Candidate A — router/expert selection behavior

Record per-token:

- router logits or normalized routing score
- top-k expert ids
- top-k margins
- selected expert weights/probabilities

This is attractive because `grok-ozempic` already measures routing fidelity.

### Candidate B — expert output summary

For one selected expert/projection, record a reduced representation such as:

- output norm
- sign/sparsity histogram
- low-dimensional random projection
- PCA/projection computed from train data only

Avoid storing full hidden vectors unless the research need is demonstrated.

### Candidate C — block-level residual correction target

Record the delta between a quantized arm and the high-precision reference:

```text
reference block output - compressed block output
```

A small student could later be tested as a learned correction module.

The issue owner should select **one** of A/B/C for v1 and document why.

## Dataset / artifact schema

Suggested JSONL/Arrow/Parquet record metadata:

```json
{
  "schema_version": 1,
  "experiment_id": "...",
  "token_index": 0,
  "block": 0,
  "arm": "fp16_reference",
  "input_ref": "...",
  "teacher": {
    "target_type": "router_topk",
    "values": {}
  },
  "provenance": {
    "grok_ozempic_commit": "...",
    "base_weight_hash": "...",
    "seed": 20260806
  }
}
```

Large arrays should use a binary/columnar artifact with JSON manifest rather than bloating JSONL.

## Provenance requirements

Every teacher artifact must record:

- exact Grok-1 source/checkpoint identity/hash used by the experiment
- block/expert/tensor identifiers
- precision arm / codec
- token/input dataset identity + hash
- seed
- `grok-ozempic` git commit
- capture code version/schema version
- whether values come from FP16/reference or a compressed arm

## Comparative capture

Where practical, capture the exact same target under:

1. FP16/reference
2. current best compressed/INT4 arm
3. relevant ternary arm

Then report target drift:

- cosine similarity where meaningful
- top-k agreement for routing targets
- MAE/RMSE for scalar/projected targets
- sign/sparsity agreement where meaningful

The purpose is to determine whether the teacher signal itself survives compression before asking a student to learn it.

## Small surrogate sanity check

Before involving an SNN, add an intentionally small baseline surrogate **outside the core quantizer path** (script/notebook/research harness is fine).

Examples:

- linear regression/classifier
- small MLP
- low-rank projection head

The surrogate should answer: is the selected teacher target learnable at all from the chosen reduced input?

If a trivial surrogate cannot beat a simple baseline, stop and document the negative result rather than escalating to SNN/FPGA complexity.

## Export contract for downstream distillation

Produce a documented artifact that can be consumed by `rmems/SynapticDistill.jl` or another trainer without importing `grok-ozempic` internals.

Suggested layout:

```text
reports/teacher-capture-<exp>/
├── manifest.json
├── targets.parquet            # or safetensors/npy if justified
├── metrics.json
├── results.md
└── README.md                   # exact reproduction command
```

## Tests

- [ ] tiny synthetic capture fixture does not require Grok weights
- [ ] manifest validates required provenance fields
- [ ] same seed/input produces stable capture metadata
- [ ] target rows retain exact block/expert/token identity
- [ ] FP16 and compressed-arm rows cannot be accidentally mixed without arm labels
- [ ] no decision logic from existing quantization experiments is changed by enabling capture
- [ ] capture can be disabled with zero material behavior change to normal experiment runs

## Acceptance criteria

This issue is complete when there is one narrow teacher-target experiment with:

1. explicit target definition,
2. reproducible capture command,
3. FP16/reference artifact,
4. at least one compressed-arm comparison,
5. target-drift metrics,
6. a small-surrogate sanity check,
7. a downstream-consumable manifest/artifact.

## Non-goals

- Exporting a full Grok-1 expert to Basys 3
- Claiming the surrogate/SNN reproduces general Grok capability
- Training Spikenaut inside `grok-ozempic`
- Adding FPGA/HDL dependencies
- Blocking current Grok-1 compression work
- Router quantization unless a separate experiment explicitly permits it

## Cross-repo follow-up

If the target is learnable and useful:

```text
Grok-1 teacher capture
        ↓
small surrogate sanity check
        ↓
SynapticDistill.jl
        ↓
Spikenaut-SNN student
        ↓
software/FPGA parity experiment
```

Related research stack epic: `rmems/Spikenaut-SNN#7`.

Suggested labels: `research`, `distillation`, `grok-1`, `neuromorphic`.

**Comment by rmems** [comment]
**Agent:** Grok Build: Grok 4.5 (xAI)

## Parked (owner decision 2026-08-13)

**Defer #84 until we actually have a created GOZ1 pack policy** worth capturing under (not just research side-table / online quant arms).

- Teacher capture remains valid as a later neuromorphic bridge (`SynapticDistill.jl` / Spikenaut).
- Do **not** start Phase 0 capture while the multi-block expert stack is still the open science question after #80 option 2.
- When unparked: prefer Candidate A (router top-k) first — harness already measures routing fidelity; keep non-goals (no FPGA, no full-expert export).

Related active path: post-#80 stacked INT4 remedies (new issue).

**Comment by rmems** [comment]
**Agent:** Grok bot: Grok-Ozempic · 2026-08-28 · cite as *Grok bot: Grok-Ozempic on #84*

Still parked. #85 / PR #89 closed option 2 at T=8192. Best compressed expert arm is still a **research INT4 side-table** (`research_int4_side` + optional LS-α), not a GOZ1 pack policy. The park condition (“wait for a created GOZ1 pack policy”) has not been met.

If this is unparked later, Candidate **A (router top-k)** is the only target the harness already measures at the full context budget:

- P1 winner b3 top-1 **0.887329** / top-2 **0.826050** (8192 tokens, seed 20260806)
- INT4-all b3 top-1 0.823730
- FP16 control b3 top-1 0.991211

That is enough to say the teacher routing signal **drifts under the best compressed arm** (about 0.063 short of the 0.95 band, and 0.104 short of FP16). Candidate C (quant vs reference residual) is the actual leftover #85 named — block-0 output drift 0.061 under α — but exporting it is a new capture path, not a pack policy.

Do not start Phase 0 from this comment. No new issue.
